tests.nixpkgs-check-by-name: Implement gradual empty arg check migration

infinisil opened the pull request:
## Context

This is part of the plan to migrate all packages to `pkgs/by-name` as detailed in https://github.com/NixOS/nixpkgs/issues/258650.

In https://github.com/NixOS/nixpkgs/pull/256792 I implemented support for the `nixpkgs-check-by-name` tool to complain when there's an unnecessary definition in `all-packages.nix` with an empty second argument like

```nix
foo = ../by-name/fo/foo/package.nix { };
```

However this was never enabled in CI with `--version v1`.
While at least in master there's no package where this applies, this is a perfect test case to make sure the tool can increase strictness over time without causing master breakages, as further detailed in https://github.com/NixOS/nixpkgs/issues/256788.

After thinking about it for too long, I realised that this is actually pretty easy to do after the changes in https://github.com/NixOS/nixpkgs/pull/261939 and https://github.com/NixOS/nixpkgs/pull/257735.

## Changes

- The `--version` flag is removed again, it was never used in CI, and it's not part of the public interface, so this is fine
- The `--base <NIXPKGS>` flag is introduced, allowing the tool to check against granular regressions. For the empty argument check this means that existing empty arguments can stay, but new ones can't be introduced.

## Example

If the base Nixpkgs has
```nix
# all-packages.nix
{
  foo = callPackage ../by-name/fo/foo/package.nix {
    withGui = true;
  };
}
```

And the main Nixpkgs has
```nix
# all-packages.nix
{
  foo = callPackage ../by-name/fo/foo/package.nix { };
}
```

Then the check fails because of the empty argument check.

But if you go the other way, the check succeeds!

## Implications

While this isn't very useful for the empty argument check specifically, the same approach can be used to also ensure that all new packages using `callPackage` use `pkgs/by-name`.

This means that we can have incremental (and automated!) PRs to do the migration, without worrying about the base branch always introducing new violations!

## Things done

- [x] Tidy code
- [x] Updated documentation
- [x] Added tests

## Follow-up PR

Once the new tooling version propagates to the nixos-unstable channel, update CI to pass the `--base` flag.

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

blaggacao commented:
Nice solution!

infinisil committed:
tests.nixpkgs-check-by-name: Minor refactor, allow more simultaneous problems

This makes it such that these two errors can both be thrown for a single
package:
- The attribute value not being a derivation
- The attribute not being a proper callPackage

The tests had to be adjusted to only throw the error they were testing
for

infinisil committed:
tests.nixpkgs-check-by-name: Intermediate refactor

This prepares the code base for the removal of the `--version` flag, to
be replaced with a flag that can specify a base version to compare the
main Nixpkgs against, in order to have gradual transitions to stricter
checks.

This refactoring does:
- Introduce the `version` module that can house the logic to increase
  strictness, with a `version::Nixpkgs` struct that contains the
  strictness conformity of a single Nixpkgs version
- Make the check return `version::Nixpkgs`
- Handle the behavior of the still-existing `--version` flag with `version::Nixpkgs`
- Introduce an intermediate `process` function to handle the top-level
  logic, especially useful in the next commit

infinisil committed:
tests.nixpkgs-check-by-name: Introduce result_map

Convenience function to run another validation over a successful validation result.

This will be usable in more locations in future commits, making the code
nicer.

infinisil committed:
tests.nixpkgs-check-by-name: Gradual migration from base Nixpkgs

This implements the option for a gradual migration to stricter checks.
For now this is only done for the check against empty non-auto-called
callPackage arguments, but in the future this can be used to ensure all
new packages make use of `pkgs/by-name`.

This is implemented by adding a `--base <BASE_NIXPKGS>` flag, which then
compares the base nixpkgs against the main nixpkgs version, making sure
that there are no regressions.

The `--version` flag is removed. While it was implemented, it was never
used in CI, so this is fine.

infinisil committed:
tests.nixpkgs-check-by-name: Test for gradual transition

This implements the ability to test gradual transitions in check
strictness, and adds one such test for the empty non-auto-called
arguments check.

philiptaron commented on pkgs/test/nixpkgs-check-by-name/src/eval.rs:
```diff
@@ -110,54 +109,67 @@ pub fn check_values(
-    Ok(validation::sequence_(package_names.iter().map(
-        |package_name| {
+    Ok(
+        validation::sequence(package_names.iter().map(|package_name| {
```
I think this `iter()` can be `into_iter()` if you want.

philiptaron commented on pkgs/test/nixpkgs-check-by-name/README.md:
```diff
@@ -8,16 +8,10 @@ This is part of the implementation of [RFC 140](https://github.com/NixOS/rfcs/pu
-- Command line: `nixpkgs-check-by-name <NIXPKGS>`
+- Command line: `nixpkgs-check-by-name [--base <BASE_NIXPKGS>] <NIXPKGS>`
 - Arguments:
+  - `<BASE_NIXPKGS>`: The path to the Nixpkgs to check against
```
If not specified, ...

infinisil replied:
Good point! Also had to fix the behavior in that case (though it doesn't matter much): https://github.com/NixOS/nixpkgs/pull/272395/commits/53b43ce0e322eb4fc8daaad8a5a597155d42379a

philiptaron commented on pkgs/test/nixpkgs-check-by-name/src/eval.rs:
```diff
@@ -39,11 +39,10 @@ enum AttributeVariant {
     package_names: Vec<String>,
-    eval_accessible_paths: Vec<&Path>,
-) -> validation::Result<()> {
+    eval_accessible_paths: &Vec<&Path>,
```
One "rusty" nit is to have the various things accepting `&Vec<Something>` accept instead the slice `&[Something]`. That's both lower power and more general. `&Vec<T>` automatically [Derefs to `&[T]`](https://doc.rust-lang.org/std/vec/struct.Vec.html#impl-Deref-for-Vec%3CT,+A%3E).

9999years replied:
It could even be `impl IntoIterator<Item = impl AsRef<Path>>` to be really general, since all you do with this value is iterate over it.

infinisil replied:
`impl` feels a bit overkill. This doesn't expose any public interface, so it should be fine for it to not be polymorphic to make it easier to read.

The `&[&Path]` in this case sounds good to me!

philiptaron commented on pkgs/test/nixpkgs-check-by-name/src/eval.rs:
```diff
@@ -110,54 +109,67 @@ pub fn check_values(
-                        } else {
-                            true
-                        };
-                        correct_file && non_empty
-                    }
-                    AttributeVariant::Other => false,
-                };
 
-                if !valid {
-                    NixpkgsProblem::WrongCallPackage {
-                        relative_package_file: relative_package_file.clone(),
-                        package_name: package_name.clone(),
+                        if correct_file {
+                            Success(version::Attribute {
+                                empty_non_auto_called: if *empty_arg {
+                                    version::EmptyNonAutoCalled::Invalid
+                                } else {
+                                    version::EmptyNonAutoCalled::Valid
+                                },
+                            })
+                        } else {
+                            NixpkgsProblem::WrongCallPackage {
+                                relative_package_file: relative_package_file.clone(),
+                                package_name: package_name.clone(),
+                            }
+                            .into()
+                        }
                     }
-                    .into()
-                } else if !attribute_info.is_derivation {
-                    NixpkgsProblem::NonDerivation {
+                    AttributeVariant::Other => NixpkgsProblem::WrongCallPackage {
                         relative_package_file: relative_package_file.clone(),
                         package_name: package_name.clone(),
                     }
-                    .into()
-                } else {
-                    Success(())
-                }
+                    .into(),
```
The current organization isn't bad at all and I would definitely accept it.

You could make a method on `NixpkgsProblem` for each condition that will be tested here, accepting whatever parameters are needed, and returning `Option<Self>` if the condition actually occured. That would make `check_values` look super sleek by sequencing together all the various checks, rather than bulky due to having the checks inlined.

9999years replied:
Yeah I would also lean towards splitting this up.

infinisil replied:
Sounds like an interesting idea, though I'd like to defer that to a future PR, since it will touch a bunch more parts of the code.

philiptaron commented on pkgs/test/nixpkgs-check-by-name/src/main.rs:
```diff
@@ -68,33 +66,25 @@ fn main() -> ExitCode {
+pub fn process<W: io::Write>(
+    base_nixpkgs: Option<&Path>,
+    main_nixpkgs: &Path,
+    eval_accessible_paths: &Vec<&Path>,
```
Same slice nit.

philiptaron commented on pkgs/test/nixpkgs-check-by-name/src/main.rs:
```diff
@@ -103,15 +93,40 @@ pub fn check_nixpkgs<W: io::Write>(
+        ))?;
+
+        if !nixpkgs_path.join(utils::BASE_SUBPATH).exists() {
+            eprintln!(
+                "Given Nixpkgs path does not contain a {} subdirectory, no check necessary.",
+                utils::BASE_SUBPATH
+            );
```
I think this wants to accept an error_writer, rather than just directly doing an `eprintln!`. The caller has one.

philiptaron commented on pkgs/test/nixpkgs-check-by-name/src/main.rs:
```diff
@@ -103,15 +93,40 @@ pub fn check_nixpkgs<W: io::Write>(
 }
 
+/// Checks whether the pkgs/by-name structure in Nixpkgs is valid,
+/// and returns to which degree it's valid for checks with increased strictness.
```
I'm not sure what this sentence is trying to communicate.

infinisil replied:
Now updated, hope this helps: https://github.com/NixOS/nixpkgs/pull/272395/commits/413dd9c03e37562c61cd89799e5eb8a88c7bb42a#diff-fc467c6e2889ff0305612175c09cd06401123d3d5e8b6cfd1e85bec22586197eL102-R104

philiptaron commented on pkgs/test/nixpkgs-check-by-name/src/main.rs:
```diff
@@ -103,15 +93,40 @@ pub fn check_nixpkgs<W: io::Write>(
+/// and returns to which degree it's valid for checks with increased strictness.
+pub fn check_nixpkgs(
+    nixpkgs_path: &Path,
+    eval_accessible_paths: &Vec<&Path>,
```
Same slice nit.

philiptaron reviewed (commented):
I did end up taking a look through the PR.  It all looks sane. I have various code organization suggestions, but the current organization isn't bad at all and I would definitely accept it. 

I haven't had a chance to run the code or tests myself, which is the only reason I'm not signing off as-is.

infinisil committed:
tests.nixpkgs-check-by-name: Fix and document behavior without --base

Previously, not passing `--base` would enforce the most strict checks.
While there's currently no actual violation of these stricter checks,
this does not match the previous behavior.

This won't matter once CI passes `--base`, the code handling the
optionality can be removed then.

infinisil committed:
tests.nixpkgs-check-by-name: Minor improvements from review

philiptaron commented on pkgs/test/nixpkgs-check-by-name/README.md:
```diff
@@ -10,8 +10,13 @@ This API may be changed over time if the CI workflow making use of it is adjuste
-  - `<NIXPKGS>`: The path to the Nixpkgs to check
+  - `<NIXPKGS>`: The path to the Nixpkgs to check.
+  - `<BASE_NIXPKGS>`: The path to the Nixpkgs to use as the base to compare `<NIXPKGS>` against.
+    This allows the strictness of checks to increase over time by only preventing _new_ violations from being introduced,
+    while allowing violations that already existed.
```
At work, we call this a [ratchet linter](https://qntm.org/ratchet).

infinisil replied:
"Ratchet" is a great term for this idea, I updated the PR to use it! https://github.com/NixOS/nixpkgs/pull/272395/commits/79618ff8cbfb69b6eb70dbc591b42cee2fed974e

infinisil commented:
Thanks for the review! Greatly appreciated as I don't really consider myself a Rust expert :heart:

I addressed all the comments, let me know if it looks good to you!

> I haven't had a chance to run the code or tests myself, which is the only reason I'm not signing off as-is.

Can confirm that the tests pass, at least on `x86_64-linux` (CI will take care of the rest). Note that if you want to run them yourself you can either `nix-build -A tests.nixpkgs-check-by-name`, or see the [development section](https://github.com/NixOS/nixpkgs/tree/master/pkgs/test/nixpkgs-check-by-name#development).

philiptaron reviewed (approved):
I'll give this an actual run later tonight.

infinisil committed:
tests.nixpkgs-check-by-name: Improve docs, introduce "ratchet" term

infinisil committed:
tests.nixpkgs-check-by-name: Move interface description into code

This would be duplicated otherwise

infinisil commented:
Finally, I moved all interface description into the `--help` output, so it's not duplicated: https://github.com/NixOS/nixpkgs/pull/272395/commits/74e8b38dbe809022d096b11b87ba33a68ba0d374

philiptaron reviewed (approved):
# 🎉

blaggacao commented on pkgs/test/nixpkgs-check-by-name/README.md:
```diff
@@ -42,9 +29,20 @@ These checks are performed by this tool:
+
+The current ratchets are:
+
+- If `pkgs.${name}` is not auto-called from `pkgs/by-name`, the `args` in its `callPackage` must not be empty,
```
That's hardly intelligible.

In part, due to that it's hard to observe an auto-called package, it's automatic after all. So how does a non-auto called look like?

I don't know what exactly is meant, but maybe could be:

> It is not allowed to introduce new `callPackage` calls in `top-level.nix`, if those would have empty `args`.

infinisil replied:
While it's not strictly necessary for users to understand these checks (that's what CI is for!), it's still good for future developers. Pushed https://github.com/NixOS/nixpkgs/pull/272395/commits/fc2d26939d2839d65553e8ca605f2440a45fc387 to improve this!

blaggacao reviewed (commented):
The description of the first ratchet could be clearer. I think it's important to prioritize intelligibility so that people intuitively understand what they should _not_ do.

infinisil committed:
tests.nixpkgs-check-by-name: Improve check clarity
